### PR TITLE
[#9382] Enable downloading of feedback session results as CSV

### PR DESCRIPTION
**Pull request by niqiukun**
Part of #9382 

**Outline of Solution**

- Migrate from `InstructorFeedbackResultsDownloadAction` to `GetSessionResultsAsCsvAction` 
- Migrate from `InstructorFeedbackResultsDownloadActionTest` to `GetSessionResultsAsCsvActionTest`
- Add basic test cases for `GetSessionResultsActionTest`
- Implement `equals()` and `hashCode()` for `SessionResultsData` for testing
- Add `downloadSessionResultsEvent` on Instructor Home Page and Instructor Sessions Page

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        gateKeeper.verifyAccessible(instructor, fs);
+    }
+
+    @Override
+    public ActionResult execute() {
+        String section = getRequestParamValue(Const.ParamsNames.SECTION_NAME);
+        String sectionDetailValue = getRequestParamValue(Const.ParamsNames.SECTION_NAME_DETAIL);
+        SectionDetail sectionDetail = SectionDetail.NOT_APPLICABLE;
+        if (section != null && sectionDetailValue != null && !sectionDetailValue.isEmpty()) {
+            if (SectionDetail.containsSectionDetail(sectionDetailValue) == null) {
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: I honestly feel `SectionDetail.containsSectionDetail` should return true/false instead. You are not even using the value it returns here.

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            sectionDetail = SectionDetail.valueOf(sectionDetailValue);
+        }
+
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+        String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+        String questionId = getRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_ID);
+        boolean isMissingResponsesShown = getBooleanRequestParamValue(
+                Const.ParamsNames.FEEDBACK_RESULTS_INDICATE_MISSING_RESPONSES);
+        boolean isStatsShown = getBooleanRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_SHOWSTATS);
+        String simulateExcessDataForTesting = getRequestParamValue("simulateExcessDataForTesting");
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: what is this supposed to be? Is it a placeholder of some kind for later?

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        String fileContent = "";
+
+        InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.id);
+
+        try {
+            if ("true".equals(simulateExcessDataForTesting)) {
+                throw new ExceedingRangeException("This session has more responses than that can be downloaded at one go.");
+            }
+
+            if (section == null || "All".equals(section)) {
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: I feel `"All".equals(section)` is a little hacky and hardcoded - any way to make it cleaner?

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if (fs == null) {
+            throw new EntityNotFoundException(new EntityDoesNotExistException("Feedback session is not found"));
+        }
+
+        InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.id);
+        gateKeeper.verifyAccessible(instructor, fs);
+    }
+
+    @Override
+    public ActionResult execute() {
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: function is a little messy - any way you can group your params that you read in from the API?

e.g. something like
```
//read in section details
-- all section related fields --

//read in course metadata

//read in options for csv

```

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAction.java [hunk 5]
```diff
@@ -25,6 +25,10 @@ protected AuthType getMinAuthLevel() {
 
     @Override
     public void checkSpecificAccessControl() {
+        if (userInfo.isAdmin) {
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/GetSessionResultsAction.java]: is this a related change?

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+import teammates.common.exception.RequestExceedingRangeException;
+import teammates.common.util.Const;
+import teammates.test.driver.CsvChecker;
+import teammates.ui.webapi.action.CsvResult;
+import teammates.ui.webapi.action.GetSessionResultsAsCsvAction;
+
+/**
+ * SUT: {@link GetSessionResultsAsCsvAction}.
+ */
+public class GetSessionResultsAsCsvActionTest extends BaseActionTest<GetSessionResultsAsCsvAction> {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: For this, what about a test for:
- typical case with section name "All"? (the if case above)
- typical case with stats shown?

File: src/web/app/pages-instructor/instructor-session-base-page.component.ts [hunk 7]
```diff
@@ -221,6 +222,29 @@ export abstract class InstructorSessionBasePageComponent {
         { courseid: model.feedbackSession.courseId, fsname: model.feedbackSession.feedbackSessionName });
   }
 
+  /**
+   * Downloads the result of a feedback session in csv.
+   */
+  downloadSessionResult(model: SessionsTableRowModel): void {
```
Comment (rrtheonlyone) [inline src/web/app/pages-instructor/instructor-session-base-page.component.ts]: for the frontend, is this the only place from which an instructor can download results? (the sessions table?)

The API has support for more params - but they are not used anywhere at the moment. Like missing responses / stats is true by default and there is no section name / detail.

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
+package teammates.test.cases.webapi;
```
Comment (rrtheonlyone) [inline src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java]: Is this test related to your PR?

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java (hunk 1 above)
Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Yes I think you are right. I'm not sure why `SectionDetail.containsSectionDetail` returns `SectionDetail` object since the value does not seem to be used anywhere else either. I will change the return type of that function.

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java (hunk 2 above)
Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: This is used in the original `InstructorFeedbackResultsDownloadAction` also to simulate the test case where the data is to large to be downloaded at one time. You may refer to https://github.com/TEAMMATES/teammates/pull/8882.

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java (hunk 3 above)
Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: This is ported from the `InstructorFeedbackResultsDownloadAction` class. I agree that this is a hardcoded value. Maybe we can remove it since `section == null` already implies all sections will be downloaded.

File: src/web/app/pages-instructor/instructor-session-base-page.component.ts (hunk 7 above)
Comment (niqiukun) [reply src/web/app/pages-instructor/instructor-session-base-page.component.ts]: For now I think this is the only place where the instructor can download session results. I'm not sure if there are other places to download results in the previous version. But we can add the feature to configure download params in future. I believe that can be a separate issue.

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java (hunk 8 above)
Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java]: Not exactly. But since I'm writing `GetSessionResultsAsCsvActionTest`, I thought it would be ok just to add `GetSessionResultsActionTest` at the same time because the test for `GetSessionResultsAction` is currently missing. There shouldn't be any breaking changes to existing code.

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAction.java (hunk 5 above)
Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAction.java]: This is because I was writing code `GetSessionResultsActionTest` which is currently missing. Although I agree that this is not exactly related to CSV downloads, but still I think the test is nevertheless necessary. And since I'm writing `GetSessionResultsAsCsvActionTest`, I thought it would be ok just to add `GetSessionResultsActionTest` at the same time.

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java (hunk 4 above)
Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Ok I have added the grouping for the params.

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java (hunk 6 above)
Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: I have added the test cases for stats hidden and shown. Thanks for spotting that! Regarding the section name "All" I think it would better to remove the hardcoded value.

Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: I discovered a problem when I tried to add test for case with stats shown. I notice that all questions in the typical data bundle are text-based questions which have no statistics to show. I am considering adding a new session to the typical data bundle, but that will break other tests that also rely on the typical data bundle. May I have some suggestions on how to resolve this issue?

Comment (rrtheonlyone) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: @xpdavid is there any way to update the typical bundle? Or is it better to just create something just for this test here?

Comment (xpdavid) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: If you don't want to create a separate json file (which in most of case is not necessary), you can manipulate/add/delete the database in `@BeforeMethod`

Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: I tried to replace the `typicalDataBundle` with a custom json file by doing the following: 
```Java
@BeforeMethod
public void setup() {
    dataBundle = loadDataBundle("/FeedbackSessionResultsCsvActionTest.json");
    removeAndRestoreDataBundle(dataBundle);
}
```
The json file has a newly added session named `Session with MCQ`. The sessions attributes can be accessed by `dataBundle.feedbackSessions.get("sessionWithMcq")`. Yet, the following exception where still thrown during the test:
```
Trying to view a non-existent feedback session: idOfTypicalCourse1/Session with MCQ
```
Did I miss anything?

File: src/web/app/pages-instructor/instructor-session-base-page.component.ts (hunk 7 above)
Comment (rrtheonlyone) [reply src/web/app/pages-instructor/instructor-session-base-page.component.ts]: okay I understand!

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java (hunk 6 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: Interesting...

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+        String questionId = getRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_ID);
+
+        // read in optional params for csv generation
+        boolean isMissingResponsesShown = getBooleanRequestParamValue(
+                Const.ParamsNames.FEEDBACK_RESULTS_INDICATE_MISSING_RESPONSES);
+        boolean isStatsShown = getBooleanRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_SHOWSTATS);
+
+        // read in params for testing
+        String simulateExcessDataForTesting = getRequestParamValue("simulateExcessDataForTesting");
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: I am still not entirely convinced by this. I feel that the excess data error should come from the logic layer which should be handled here. 

Why is it part of our API params?

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    protected String getRequestMethod() {
+        return GET;
+    }
+
+    @Override
+    @Test
+    protected void testExecute() throws Exception {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java]: Where are the tests for student's accessing their results?

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+    protected String getActionUri() {
+        return Const.ResourceURIs.RESULT_CSV;
+    }
+
+    @Override
+    protected String getRequestMethod() {
+        return GET;
+    }
+
+    private void loadCustomBundle() {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java]: I see what you are doing here and it works! Although its a little different from how our other tests work.. maybe @xpdavid can help improve this.

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java (hunk 9 above)
Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: The excess data error indeed comes from the logic layer in practice. However for testing it is not practical to generate large amount of data just to trigger the error. Hence, this param is kept there so that the error is triggered manually and exception handling can be tested. But you are right in the way that it does not actually test whether the exception will be thrown at the right time, so it is only tested partially. I'm not really sure if there are better ways to do this.

Comment (rrtheonlyone) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: For just testing this on the API layer, you could just mock the logic layer call to throw the relevant exception. For E2E testing, we probably HAVE to generate big data! At least that's what I think :) 

It seems a little odd that we are adding a functionality just for our own internal testing. However, we def have to discuss on this with the rest and see.

Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: > For just testing this on the API layer, you can mock the logic layer call to throw the relevant exception.

Sorry but I don't quite get this part. Could you explain a bit more? Thanks :)

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java (hunk 10 above)
Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsActionTest.java]: I have added them in the latest commit. Thanks for spotting that!

File: src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java (hunk 9 above)
Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Adding this line in production code just for testing?

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Who will throw `ExceedingRangeException`?

Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: > Adding this line in production code just for testing?

This is carried forward from https://github.com/TEAMMATES/teammates/pull/8882. This is how the previous developer describes its use:

```
In the `InstructorFeedbackResultsDownloadActionTest` we add a test
where we put in an extra parameter `largeData=true` and then in the Action class
we put an if condition to see if this parameter is not null and true and if it is then we
throw the required exception for the `code` of exception part to be called and assert
in the test if thats what we expect to happen
```

I would like to have some suggestions on improving this if this is not a good programming practice.

> Who will throw `ExceedingRangeException`?

`logic.getFeedbackSessionResultSummaryAsCsv(...)` and `logic.getFeedbackSessionResultSummaryInSectionAsCsv(...)` throw this exception.

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Then we should meet the condition to let `getFeedbackSessionResultSummaryAsCsv` or `getFeedbackSessionResultSummaryInSectionAsCsv` throw the exception? Or even remove the test because it actually test nothing.

Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: > Then we should meet the condition to let `getFeedbackSessionResultSummaryAsCsv` or `getFeedbackSessionResultSummaryInSectionAsCsv` throw the exception? Or even remove the test because it actually test nothing.

The test file is not large enough to trigger the condition that throws this exception. I think it may be better to remove this test then.

Comment (rrtheonlyone) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Hi @niqiukun , sorry missed your earlier question, I was thinking more of something like this:
https://se-education.org/se-book/testing/#stubs

You can just mock the logic layer to throw the relevant exception for testing the API layer. 
For E2E testing, we can just generate a large csv into the database.

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java]: Introducing mock for logic layer requires a huge effort without help from mocking framework. We might want to do it separately.  I think we can just remove the test till now.

File: src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java (hunk 11 above)
Comment (xpdavid) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java:34]: I think it is OK to create a new data file for this test case ad download as CSV is a big feature. One thing to note here is that please override the corresponding method to avoid `typicalDataBundle` from loading.

Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java:34]: The way I did this was only using the custom data bundle for session with statistics. By doing that I don't have to duplicate the typical data bundle for those test cases that do not require the custom data bundle. And that's why `typicalDataBundle` is still being loaded here.

Comment (xpdavid) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java:34]: So it your `json` file is incremental based on `typicalDataBundle`?

Comment (niqiukun) [reply src/test/java/teammates/test/cases/webapi/GetSessionResultsAsCsvActionTest.java:34]: It only contains one session but with a MCQ question instead of a text question, so that there will be statistics that can be downloaded. I did not include those sessions that were originally in the `typicalDataBundle`. But this `json` file is only used for two test cases: `Typical case: results with statistics hidden` and `Typical case: results with statistics shown`. All other test cases use `typicalDataBundle` instead.

File: src/main/java/teammates/ui/webapi/output/SessionResultsData.java [hunk 12]
```diff
@@ -77,6 +78,30 @@ public SessionResultsData(FeedbackSessionResultsBundle bundle, StudentAttributes
         return questions;
     }
 
+    @Override
+    public boolean equals(Object obj) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/output/SessionResultsData.java:82]: Is this method added just for the sake of testing?

If yes, please convert it to helper method in the test file or you are responsible to write a comprehensive test for it (which I think it is not necessary).

Comment (niqiukun) [reply src/main/java/teammates/ui/webapi/output/SessionResultsData.java:82]: I see your point. I would like to convert it to a helper method, but here `QuestionOutput` is an inner class that has private access. Should I change the accessibility of the class or is there any wordaround?

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/output/SessionResultsData.java:82]: You can make the change.
